pyo3-macros-backend/src/pyclass.rs

```diff
@@ -440,24 +476,127 @@ impl<'a> PyClassEnum<'a> {
     }
 }
 
-pub fn build_py_enum(
-    enum_: &mut syn::ItemEnum,
-    mut args: PyClassArgs,
-    method_type: PyClassMethodsType,
-) -> syn::Result<TokenStream> {
-    args.options.take_pyo3_options(&mut enum_.attrs)?;
+struct PyClassComplexEnum<'a> {
+    ident: &'a syn::Ident,
+    variants: Vec<PyClassEnumVariant<'a>>,
+}
 
-    if let Some(extends) = &args.options.extends {
-        bail_spanned!(extends.span() => "enums can't extend from other classes");
-    } else if let Some(subclass) = &args.options.subclass {
-        bail_spanned!(subclass.span() => "enums can't be inherited by other classes");
-    } else if enum_.variants.is_empty() {
-        bail_spanned!(enum_.brace_token.span.join() => "#[pyclass] can't be used on enums without any variants");
+impl<'a> PyClassComplexEnum<'a> {
+    fn new(enum_: &'a mut syn::ItemEnum) -> syn::Result<Self> {
+        fn extract_variant_data(variant: &mut syn::Variant) -> syn::Result<PyClassEnumVariant<'_>> {
+            use syn::Fields;
+            let ident = &variant.ident;
+            let options = EnumVariantPyO3Options::take_pyo3_options(&mut variant.attrs)?;
+
+            let variant = match &variant.fields {
+                Fields::Unit => {
+                    bail_spanned!(variant.span() => "Unit variants are not yet supported in complex enums.")
+                }
+                Fields::Named(fields) => {
+                    let fields = fields
+                        .named
+                        .iter()
+                        .map(|field| PyClassEnumVariantNamedField {
+                            ident: field.ident.as_ref().expect("named field has an identifier"),
+                            ty: &field.ty,
+                            span: field.span(),
+                        })
+                        .collect();
+
+                    PyClassEnumVariant::Struct(PyClassEnumStructVariant {
+                        ident,
+                        fields,
+                        options,
+                    })
+                }
+                Fields::Unnamed(_) => {
+                    bail_spanned!(variant.span() => "Tuple variants are not yet supported in complex enums.")
+                }
+            };
+
+            Ok(variant)
+        }
+
+        let ident = &enum_.ident;
+
+        let variants: Vec<_> = enum_
+            .variants
+            .iter_mut()
+            .map(extract_variant_data)
+            .collect::<syn::Result<_>>()?;
+
+        Ok(Self { ident, variants })
     }
+}
 
-    let doc = utils::get_doc(&enum_.attrs, None);
-    let enum_ = PyClassEnum::new(enum_)?;
-    impl_enum(enum_, &args, doc, method_type)
+enum PyClassEnumVariant<'a> {
+    // TODO(mkovaxx): Unit(PyClassEnumUnitVariant<'a>),
+    Struct(PyClassEnumStructVariant<'a>),
+    // TODO(mkovaxx): Tuple(PyClassEnumTupleVariant<'a>),
+}
```

Comment:
@davidhewitt I've introduced this type to make it easier to find where to start adding unit and tuple variant support.